File: src/Tests/Behavioral/StringLenUtf8Bytes/main.go
package main

import "fmt"

// Exercises len(string) over multi-byte literals: Go counts UTF-8 BYTES, so the values
// below differ from the UTF-16 char counts a naive C# string.Length would report. The
// table shape keeps the literals as plain C# strings in the emission (the u8-suppressed
// argument path), which is exactly the overload under guard. Printed output stays ASCII.
var tests = []struct {
	name string
	n    int
}{
	{"ascii", len("abc")},
	{"bmp2", len("aΩb")},        // Ω: 2-byte UTF-8, 1 UTF-16 char
	{"bmp3", len("a☺b☻c")}, // ☺☻: 3-byte UTF-8, 1 UTF-16 char each
	{"astral", len("z\U0001d56b")},   // 𝕫: 4-byte UTF-8, UTF-16 surrogate pair
}

func main() {
	for _, t := range tests {
		fmt.Println(t.name, t.n)
	}
	fmt.Println(len("a☺b☻"), len("héllo"), len("日本語"))
}

File: src/Tests/Behavioral/StringLenUtf8Bytes/main.cs
namespace go;

using fmt = fmt_package;

partial class main_package {


[GoType("dyn")] partial struct testsᴛ1 {
    internal @string name;
    internal nint n;
}
internal static slice<testsᴛ1> tests = new testsᴛ1[]{
    new("ascii"u8, len("abc")),
    new("bmp2"u8, len("aΩb")),
    new("bmp3"u8, len("a☺b☻c")),
    new("astral"u8, len("z\U0001d56b"))
}.slice();

internal static void Main() {
    foreach (var (_, t) in tests) {
        fmt.Println(t.name, t.n);
    }
    fmt.Println(len("a☺b☻"), len("héllo"), len("日本語"));
}

} // end main_package
